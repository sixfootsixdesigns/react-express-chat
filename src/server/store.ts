import * as shortid from 'shortid';

export interface IMessage {
  author: IUser;
  message: string;
  id: string;
}

export interface IUser {
  name: string;
  id: string;
}

const MAX_MESSAGES = 50;
let users: IUser[] = [];
let messages: IMessage[] = [];

const addMessage = (author: IUser, message: string) => {
  const data: IMessage = {
    author: author,
    message: message,
    id: shortid.generate()
  };
  messages.push(data);

  messages = messages.slice(MAX_MESSAGES * -1);

  return data;
};

const dropUser = (remove: IUser) => {
  users = users.filter(user => user.id !== remove.id);
};

const addUser = (username: string) => {
  const user: IUser = {
    name: username,
    id: shortid.generate()
  };
  users.push(user);
  return user;
};

const getUsers = () => users;
const getMessages = () => messages;

const reset = () => {
  users = [];
  messages = [];
};

export const storeAPI = {
  MAX_MESSAGES,
  addUser,
  getUsers,
  getMessages,
  addMessage,
  reset,
  dropUser
};
