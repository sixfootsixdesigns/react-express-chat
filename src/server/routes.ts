import { Router } from 'express';
import { storeAPI } from './store';
import socket from 'socket.io';

export const routes = (socket: socket.Server) => {
  const router = Router();

  router
    .get('/messages', (req: Request, res: any) => {
      res.json(storeAPI.getMessages());
    })

    .post('/add-message', (req: Request, res: any) => {
      const body: any = req.body;
      const message = storeAPI.addMessage(body.author, body.message);

      socket.emit('NewMessage', storeAPI.getMessages());

      res.status(201);
      res.json(message);
    })

    .get('/users', (req: Request, res: any) => {
      res.json(storeAPI.getUsers());
    })

    .post('/add-user', (req: Request, res: any) => {
      const body: any = req.body;
      const user = storeAPI.addUser(body.username);

      socket.emit('UserAdded', storeAPI.getUsers());

      res.status(201);
      res.json(user);
    });

  return router;
};
