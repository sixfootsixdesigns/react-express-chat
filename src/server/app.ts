import express from 'express';
import * as bodyParser from 'body-parser';
import { routes } from './routes';
import * as http from 'http';
import socket from 'socket.io';
import { storeAPI, IUser } from './store';

const app = express();

const appServer = http.createServer(app);
const socketIO = socket(appServer);

app
  .use(bodyParser.json())
  .use(routes(socketIO))
  .use(express.static('build'));

socketIO.on('connection', socket => {
  socket.on('disconnecting', (user: IUser) => {
    storeAPI.dropUser(user);
    socketIO.emit('UserLeft', storeAPI.getUsers());
  });
});

export { appServer, socketIO };
