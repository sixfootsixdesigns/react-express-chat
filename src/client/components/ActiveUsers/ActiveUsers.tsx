import React from 'react';
import { IUser } from '../../../server/store';

interface ActiveUsersProps {
  users: IUser[];
  currentUser?: IUser;
  loading?: boolean;
}

export class ActiveUsers extends React.Component<ActiveUsersProps> {
  public render() {
    const { users, currentUser, loading } = this.props;

    if (loading) {
      return (
        <div className="ActiveUsers">
          <h2>Users</h2>
          loading...
        </div>
      );
    }

    return (
      <div className="ActiveUsers">
        <h2>Users</h2>
        {users.length ? (
          <ul>
            {users.map(user => (
              <li key={user.id}>
                {currentUser && currentUser.id === user.id && '** '}
                {user.name}
              </li>
            ))}
          </ul>
        ) : (
          <p>No Active Users</p>
        )}
      </div>
    );
  }
}
