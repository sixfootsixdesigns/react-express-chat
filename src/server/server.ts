import { appServer } from './app';

const port = Number(process.env.PORT || 4001);

appServer.listen(port, () => {
  console.log(`listening on port: ${port}`);
});
