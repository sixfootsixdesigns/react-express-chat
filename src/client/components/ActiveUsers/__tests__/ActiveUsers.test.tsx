import * as React from 'react';
import { render, cleanup } from 'react-testing-library';
import { ActiveUsers } from '../ActiveUsers';
import { IUser } from '../../../../server/store';

describe('Active Users Component', () => {
  afterEach(() => {
    cleanup();
  });

  it('should render with users and highlight current user', () => {
    const users: IUser[] = [
      {
        name: 'bob',
        id: 'abc'
      },
      {
        name: 'dave',
        id: 'aslfj'
      }
    ];
    const currUser = users[1];

    const { container } = render(
      <ActiveUsers currentUser={currUser} users={users} />
    );

    expect(container.firstChild).toMatchSnapshot();
  });

  it('should render without users', () => {
    const users: IUser[] = [];

    const { container } = render(<ActiveUsers users={users} />);

    expect(container.firstChild).toMatchSnapshot();
  });

  it('should show loading state', () => {
    const users: IUser[] = [];

    const { container } = render(<ActiveUsers loading={true} users={users} />);

    expect(container.firstChild).toMatchSnapshot();
  });
});
