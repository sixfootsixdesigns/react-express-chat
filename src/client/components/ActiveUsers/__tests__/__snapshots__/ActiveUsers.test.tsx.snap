// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Active Users Component should render with users and highlight current user 1`] = `
<div
  class="ActiveUsers"
>
  <h2>
    Users
  </h2>
  <ul>
    <li>
      bob
    </li>
    <li>
      ** 
      dave
    </li>
  </ul>
</div>
`;

exports[`Active Users Component should render without users 1`] = `
<div
  class="ActiveUsers"
>
  <h2>
    Users
  </h2>
  <p>
    No Active Users
  </p>
</div>
`;

exports[`Active Users Component should show loading state 1`] = `
<div
  class="ActiveUsers"
>
  <h2>
    Users
  </h2>
  loading...
</div>
`;
